[WIP] PR for - Support Solidity's library syntax #32

anurag-git opened the pull request:
Fixed the issues and following Scenarios tested
=======================================
1. library declared in same file a contract
2. library in different file but same directory and imported as 
	import "./SafeMath.sol"
3. library in different file and different directory and imported as 
	import "./lib/SafeMath.sol";

Would like to know any other scenarios and fixes to be done for this issue.

anurag-git commented:
Fixes #32

GNSPS commented:
One thing that comes to mind is that for the helper I could envision something like the contract profiler that caches info about libraries in a file and then that info is available when visiting `FunctionCall` nodes to check if the member access is a library function being used.

So basically a "library profiler" of sorts.

anurag-git commented:
I kind of doing something similar. Added all required functions uaed in
ftrace in contract profiler but facing some issue while FunctionCall is
called. It exits after first loop, instead of looping for all functions,
modifiers etc.

Any clues to procide?

Regards,
Anurag Dashputre

On Fri 31 Aug, 2018 22:09 Gonçalo Sá, <notifications@github.com> wrote:

> One thing that comes to mind is that for the helper I could envision
> something like the contract profiler that caches info about libraries in a
> file and then that info is available when visiting FunctionCall nodes to
> check if the member access is a library function being used.
>
> So basically a "library profiler" of sorts.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ConsenSys/surya/pull/41#issuecomment-417721763>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AbZ-60h6wCj_lgZNoCSc9NM3ZssqPt5iks5uWWbHgaJpZM4WM3aL>
> .
>

anurag-git commented:
Fix details
==================
ftrace.js
---------
Updated to move `ImportDirective` and `UsingForDeclaration` to `contractProfiler.js`.

contractProfiler.js
---------------------
Added all the function definitions used in `ftrace.js` including `ImportDirective` and `UsingForDeclaration`.

test.sh
--------
Added 9 test cases as mentioned below.

test/contracts folder
-----------------------
Added contracts and libraries needed for all the 9 test cases. 

Tests Automated
====================
1. library defined same file as contract file
2. library imported from different file in same location as contract
3. library imported from different file in different location as contract
4. library in same file as contract file + 1 import of another contract
5. library imported from different file + 1 import of another contract
6. contract function having 2 different library functions
7. contract function having 3 different library functions
8. contract function without external library functions
9. contract function importing 2 different libraries

Steps to execute automated test cases
=======================================
npm test ftrace
 @GNSPS , please review and provide your feedback

anurag-git commented:
Fixes #32

GNSPS commented:
This all looks good to me! I'll accept the PR. 😄

GNSPS commented on src/utils/contractProfiler.js:
```diff
@@ -65,14 +90,104 @@ function contractProfilesFromFile(file) {
+      varProfile = new Object()
+    },
+
+// TODO:Need to re-check this functionlity
```
What does this comment mean exactly?

anurag-git replied:
Actually this code is not working, hence comment. Removing it in next PR

GNSPS commented:
Another thing that is not working is when path redirects are being used in the compiler. Like when using Open Zeppelin's libraries installed from NPM (which happens most of the times). 

Relevant for this issue would be this part of Truffle's documentation: https://truffleframework.com/docs/truffle/getting-started/package-management-via-npm#within-your-contracts

I feel that this could be resolved fairly easily and after this fix and resolving the comments of the review I'm ready to approve in Gitcoin.

GNSPS commented:
What I meant with the comment above is that we should support Truffle's import behavior by default with some flag.

anurag-git commented on test/test.sh:
```diff
@@ -3,8 +3,31 @@
+elif [ "$commandName" == "ftrace" ]; then
+	echo "Test 1: Library defined in same file as contract file"
+	node ./bin/surya ${commandName} SomeContract::doSomething all ./test/contracts/Test1.sol
+	echo "Test 2: Lbrary imported from different file in same location as contract"
```
> There's a typo there. Could you change it to `Library`, please?

Done

GNSPS commented:
It still breaks for me in some codebases but I feel the work is considered enough based on my loose specifications, though. So I'll just accept it and work on cleaning it up myself.

So, I'll create a new branch called `feat/library-support` and you can point the PR to that.

GNSPS commented on src/utils/contractProfiler.js:
```diff
@@ -106,10 +108,28 @@ function contractProfilesFromFile(file) {
+      }
+
+      let [dirName,restDir] = path.dirname(node.path).split('/', 2)
+      let lib=''
```
Also this variable naming is not quite right, since it can be anything and not only a library. Maybe something like `let importPath` would be better

GNSPS commented on src/utils/contractProfiler.js:
```diff
@@ -65,14 +92,111 @@ function contractProfilesFromFile(file) {
+        content = fs.readFileSync(lib).toString('utf-8')
+      } catch (e) {
+        if (e.code === 'EISDIR') {
+          console.error(`Skipping directory ${lib}`)
```
Also no need to skip folders here since no one should be importing folders and that is disallowed in the compiler. We should just error out.

anurag-git commented:
> It still breaks for me in some codebases but I feel the work is considered enough based on my loose specifications, though. So I'll just accept it and work on cleaning it up myself.
> 
> So, I'll create a new branch called `feat/library-support` and you can point the PR to that.

Thanks @GNSPS

GNSPS commented:
Please point the PR to the `library-support` branch so I can merge it. 😄

anurag-git commented:
Closing as new PR #47 created to further track this request
